apollo-router/src/plugins/record_replay/replay.rs

```diff
@@ -0,0 +1,418 @@
+            .and_operation_name(client_request.operation_name.clone())
+            .variables(client_request.variables.clone())
+            .headers(request_headers)
+            .context(Context::default())
+            .uri(Uri::from_static("http://localhost:8080"))
+            .method(Method::POST)
+            .build()?;
+
+        Ok(req.try_into()?)
+    }
+
+    pub(crate) fn supergraph_sdl(&self) -> String {
+        self.recording.supergraph_sdl.clone()
+    }
+}
+
+#[async_trait::async_trait]
+impl Plugin for Replay {
+    type Config = ();
+
+    async fn new(_init: PluginInit<Self::Config>) -> Result<Self, BoxError> {
+        unreachable!()
+    }
+
+    fn router_service(&self, service: router::BoxService) -> router::BoxService {
+        let report = self.report.clone();
+        let recorded_headers = self.recording.client_response.headers.clone();
+
+        ServiceBuilder::new()
+            .map_response(move |res: router::Response| {
+                // TODO - check matching headers?
+                for (k, recorded_values) in recorded_headers.iter() {
+                    let recorded_set = recorded_values
+                        .iter()
+                        .map(|v| v.to_string())
+                        .collect::<HashSet<_>>();
+                    let runtime_values = res
+                        .response
+                        .headers()
+                        .get_all(k)
+                        .iter()
+                        .map(|v| String::from(v.to_str().unwrap()))
+                        .collect::<HashSet<_>>();
+
+                    let missing_values =
+                        recorded_set.difference(&runtime_values).collect::<Vec<_>>();
+
+                    if !missing_values.is_empty() {
+                        report.lock().unwrap().push(ReplayReport::HeaderDifference {
+                            name: k.clone(),
+                            recorded: recorded_values.clone(),
+                            runtime: runtime_values.iter().map(|v| v.to_string()).collect(),
+                        });
+                    }
+                }
+
+                res
+            })
+            .service(service)
+            .boxed()
+    }
+
+    fn supergraph_service(&self, service: supergraph::BoxService) -> supergraph::BoxService {
+        let report = self.report.clone();
+        let recorded_chunks = self.recording.client_response.chunks.clone();
+
+        ServiceBuilder::new()
+            .map_request(|req: supergraph::Request| {
+                // TODO - nothing? request is coming from elsewhere
+                req
+            })
+            .map_response(|res: supergraph::Response| {
+                let mut i = 0;
+                res.map_stream(move |chunk| {
+                    let recorded_chunk = &recorded_chunks[i];
+                    let chunk = chunk.clone();
+
+                    // TODO - json string equality is sufficient?
```

Comment:
I don't know, do we have cases when we have an HashMap in json data ? That might be annoying because depending on the machine it's running it might have different orders. We will discover it I guess